<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>blockForest — blockForest • blockForest</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="blockForest — blockForest"><meta property="og:description" content="Block forests without parameter tuning.
Use blockfor for standard interface.
This function is called by blockfor
and will rarely be considered directly by the user (since parameter tuning
is required in applications)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">blockForest</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/bips-hb/blockForest/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>blockForest</h1>
    <small class="dont-index">Source: <a href="https://github.com/bips-hb/blockForest/blob/master/R/blockForest.R" class="external-link"><code>R/blockForest.R</code></a></small>
    <div class="hidden name"><code>blockForest.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Block forests without parameter tuning.
Use <code><a href="blockfor.html">blockfor</a></code> for standard interface.
This function is called by <code><a href="blockfor.html">blockfor</a></code>
and will rarely be considered directly by the user (since parameter tuning
is required in applications).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">blockForest</span><span class="op">(</span></span>
<span>  formula <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  num.trees <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  mtry <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  importance <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  write.forest <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  probability <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  min.node.size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  replace <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sample.fraction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">replace</span>, <span class="fl">1</span>, <span class="fl">0.632</span><span class="op">)</span>,</span>
<span>  case.weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitrule <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  num.random.splits <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  minprop <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  split.select.weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  always.split.variables <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  blocks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  block.method <span class="op">=</span> <span class="st">"BlockForest"</span>,</span>
<span>  block.weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  respect.unordered.factors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scale.permutation.importance <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  keep.inbag <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  holdout <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  quantreg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  num.threads <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  save.memory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dependent.variable.name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  status.variable.name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  classification <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-formula">formula<a class="anchor" aria-label="anchor" href="#arg-formula"></a></dt>
<dd><p>Object of class <code>formula</code> or <code>character</code> describing the model to fit. Interaction terms supported only for numerical variables.</p></dd>


<dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>Training data of class <code>data.frame</code>, <code>matrix</code>, <code>dgCMatrix</code> (Matrix) or <code>gwaa.data</code> (GenABEL).</p></dd>


<dt id="arg-num-trees">num.trees<a class="anchor" aria-label="anchor" href="#arg-num-trees"></a></dt>
<dd><p>Number of trees.</p></dd>


<dt id="arg-mtry">mtry<a class="anchor" aria-label="anchor" href="#arg-mtry"></a></dt>
<dd><p>This is either a number specifying the number of variables sampled for each
split from all variables (for variants "VarProb" and "SplitWeights")
or a vector of length equal to the number of blocks, where the m-th entry of the
vector gives the number of variables to sample from block m (for variants "BlockForest", "RandomBlock", and "BlockVarSel").
The default values are sqrt(p_1) + sqrt(p_2) + ... sqrt(p_M) and (sqrt(p_1), sqrt(p_2), ..., sqrt(p_M)), respectively,
where p_m denotes the number of variables in the m-th block (m = 1, ..., M) and sqrt() denoted the square root function.</p></dd>


<dt id="arg-importance">importance<a class="anchor" aria-label="anchor" href="#arg-importance"></a></dt>
<dd><p>Variable importance mode, one of 'none', 'impurity', 'impurity_corrected', 'permutation'. The 'impurity' measure is the Gini index for classification, the variance of the responses for regression and the sum of test statistics (see <code>splitrule</code>) for survival.</p></dd>


<dt id="arg-write-forest">write.forest<a class="anchor" aria-label="anchor" href="#arg-write-forest"></a></dt>
<dd><p>Save <code>blockForest.forest</code> object, required for prediction. Set to <code>FALSE</code> to reduce memory usage if no prediction intended.</p></dd>


<dt id="arg-probability">probability<a class="anchor" aria-label="anchor" href="#arg-probability"></a></dt>
<dd><p>Grow a probability forest as in Malley et al. (2012).</p></dd>


<dt id="arg-min-node-size">min.node.size<a class="anchor" aria-label="anchor" href="#arg-min-node-size"></a></dt>
<dd><p>Minimal node size. Default 1 for classification, 5 for regression, 3 for survival, and 10 for probability.</p></dd>


<dt id="arg-replace">replace<a class="anchor" aria-label="anchor" href="#arg-replace"></a></dt>
<dd><p>Sample with replacement.</p></dd>


<dt id="arg-sample-fraction">sample.fraction<a class="anchor" aria-label="anchor" href="#arg-sample-fraction"></a></dt>
<dd><p>Fraction of observations to sample. Default is 1 for sampling with replacement and 0.632 for sampling without replacement. For classification, this can be a vector of class-specific values.</p></dd>


<dt id="arg-case-weights">case.weights<a class="anchor" aria-label="anchor" href="#arg-case-weights"></a></dt>
<dd><p>Weights for sampling of training observations. Observations with larger weights will be selected with higher probability in the bootstrap (or subsampled) samples for the trees.</p></dd>


<dt id="arg-splitrule">splitrule<a class="anchor" aria-label="anchor" href="#arg-splitrule"></a></dt>
<dd><p>Splitting rule, default "extratrees". Other options are "gini" for classification and probability estimation, "variance", or "maxstat" for regression and "logrank", "C" or "maxstat" for survival.</p></dd>


<dt id="arg-num-random-splits">num.random.splits<a class="anchor" aria-label="anchor" href="#arg-num-random-splits"></a></dt>
<dd><p>For "extratrees" splitrule.: Number of random splits to consider for each candidate splitting variable.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>For "maxstat" splitrule: Significance threshold to allow splitting.</p></dd>


<dt id="arg-minprop">minprop<a class="anchor" aria-label="anchor" href="#arg-minprop"></a></dt>
<dd><p>For "maxstat" splitrule: Lower quantile of covariate distribution to be considered for splitting.</p></dd>


<dt id="arg-split-select-weights">split.select.weights<a class="anchor" aria-label="anchor" href="#arg-split-select-weights"></a></dt>
<dd><p>Numeric vector with weights between 0 and 1, representing the probability to select variables for splitting. Alternatively, a list of size num.trees, containing split select weight vectors for each tree can be used. Use this for the "VarProb" variant.</p></dd>


<dt id="arg-always-split-variables">always.split.variables<a class="anchor" aria-label="anchor" href="#arg-always-split-variables"></a></dt>
<dd><p>Character vector with variable names to be always selected in addition to the <code>mtry</code> variables tried for splitting.</p></dd>


<dt id="arg-blocks">blocks<a class="anchor" aria-label="anchor" href="#arg-blocks"></a></dt>
<dd><p>Block memberships of the variables. See <code><a href="blockfor.html">blockfor</a></code> for details.</p></dd>


<dt id="arg-block-method">block.method<a class="anchor" aria-label="anchor" href="#arg-block-method"></a></dt>
<dd><p>Variant to use. Options are: "BlockForest" (default), "RandomBlock", "BlockVarSel", "SplitWeights".</p></dd>


<dt id="arg-block-weights">block.weights<a class="anchor" aria-label="anchor" href="#arg-block-weights"></a></dt>
<dd><p>Tuning parameter values for the blocks in the variants. A vector of length equal to the number of blocks or a list with vectors containing tree-wise values. For block.method='RandomBlock' these are the block sample probabilities.</p></dd>


<dt id="arg-respect-unordered-factors">respect.unordered.factors<a class="anchor" aria-label="anchor" href="#arg-respect-unordered-factors"></a></dt>
<dd><p>Handling of unordered factor covariates. One of 'ignore', 'order' and 'partition'. For the "extratrees" splitrule the default is "partition" for all other splitrules 'ignore'. Alternatively TRUE (='order') or FALSE (='ignore') can be used. See below for details.</p></dd>


<dt id="arg-scale-permutation-importance">scale.permutation.importance<a class="anchor" aria-label="anchor" href="#arg-scale-permutation-importance"></a></dt>
<dd><p>Scale permutation importance by standard error as in (Breiman 2001). Only applicable if permutation variable importance mode selected.</p></dd>


<dt id="arg-keep-inbag">keep.inbag<a class="anchor" aria-label="anchor" href="#arg-keep-inbag"></a></dt>
<dd><p>Save how often observations are in-bag in each tree.</p></dd>


<dt id="arg-holdout">holdout<a class="anchor" aria-label="anchor" href="#arg-holdout"></a></dt>
<dd><p>Hold-out mode. Hold-out all samples with case weight 0 and use these for variable importance and prediction error.</p></dd>


<dt id="arg-quantreg">quantreg<a class="anchor" aria-label="anchor" href="#arg-quantreg"></a></dt>
<dd><p>Prepare quantile prediction as in quantile regression forests (Meinshausen 2006). Regression only. Set <code>keep.inbag = TRUE</code> to prepare out-of-bag quantile prediction.</p></dd>


<dt id="arg-num-threads">num.threads<a class="anchor" aria-label="anchor" href="#arg-num-threads"></a></dt>
<dd><p>Number of threads. Default is number of CPUs available.</p></dd>


<dt id="arg-save-memory">save.memory<a class="anchor" aria-label="anchor" href="#arg-save-memory"></a></dt>
<dd><p>Use memory saving (but slower) splitting mode. No effect for survival and GWAS data. Warning: This option slows down the tree growing, use only if you encounter memory problems.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Show computation status and estimated runtime.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Random seed. Default is <code>NULL</code>, which generates the seed from <code>R</code>. Set to <code>0</code> to ignore the <code>R</code> seed.</p></dd>


<dt id="arg-dependent-variable-name">dependent.variable.name<a class="anchor" aria-label="anchor" href="#arg-dependent-variable-name"></a></dt>
<dd><p>Name of dependent variable, needed if no formula given. For survival forests this is the time variable.</p></dd>


<dt id="arg-status-variable-name">status.variable.name<a class="anchor" aria-label="anchor" href="#arg-status-variable-name"></a></dt>
<dd><p>Name of status variable, only applicable to survival data and needed if no formula given. Use 1 for event and 0 for censoring.</p></dd>


<dt id="arg-classification">classification<a class="anchor" aria-label="anchor" href="#arg-classification"></a></dt>
<dd><p>Only needed if data is a matrix. Set to <code>TRUE</code> to grow a classification forest.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Object of class <code>blockForest</code> with elements</p>
<dl><dt><code>forest</code></dt>
<dd><p>Saved forest (If write.forest set to TRUE). Note that the variable IDs in the <code>split.varIDs</code> object do not necessarily represent the column number in R.</p></dd>

  <dt><code>predictions</code></dt>
<dd><p>Predicted classes/values, based on out of bag samples (classification and regression only).</p></dd>

  <dt><code>variable.importance</code></dt>
<dd><p>Variable importance for each independent variable.</p></dd>

  <dt><code>prediction.error</code></dt>
<dd><p>Overall out of bag prediction error. For classification this is the fraction of missclassified samples, for probability estimation and regression the mean squared error and for survival one minus Harrell's C-index.</p></dd>

  <dt><code>r.squared</code></dt>
<dd><p>R squared. Also called explained variance or coefficient of determination (regression only). Computed on out of bag data.</p></dd>

  <dt><code>confusion.matrix</code></dt>
<dd><p>Contingency table for classes and predictions based on out of bag samples (classification only).</p></dd>

  <dt><code>unique.death.times</code></dt>
<dd><p>Unique death times (survival only).</p></dd>

  <dt><code>chf</code></dt>
<dd><p>Estimated cumulative hazard function for each sample (survival only).</p></dd>

  <dt><code>survival</code></dt>
<dd><p>Estimated survival function for each sample (survival only).</p></dd>

  <dt><code>call</code></dt>
<dd><p>Function call.</p></dd>

  <dt><code>num.trees</code></dt>
<dd><p>Number of trees.</p></dd>

  <dt><code>num.independent.variables</code></dt>
<dd><p>Number of independent variables.</p></dd>

  <dt><code>mtry</code></dt>
<dd><p>Value of mtry used.</p></dd>

  <dt><code>min.node.size</code></dt>
<dd><p>Value of minimal node size used.</p></dd>

  <dt><code>treetype</code></dt>
<dd><p>Type of forest/tree. classification, regression or survival.</p></dd>

  <dt><code>importance.mode</code></dt>
<dd><p>Importance mode used.</p></dd>

  <dt><code>num.samples</code></dt>
<dd><p>Number of samples.</p></dd>

  <dt><code>inbag.counts</code></dt>
<dd><p>Number of times the observations are in-bag in the trees.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>See <code><a href="blockfor.html">blockfor</a></code> and the <code>ranger</code> package.</p>
    </div>
    <div id="references">
    <h2>References</h2>

<ul><li><p>Hornung, R. &amp; Wright, M. N. (2019) Block Forests: random forests for blocks of clinical and omics covariate data. BMC Bioinformatics 20:358. <a href="https://doi.org/10.1186/s12859-019-2942-y" class="external-link">doi:10.1186/s12859-019-2942-y</a>
.</p></li>
<li><p>Wright, M. N. &amp; Ziegler, A. (2017). ranger: A Fast Implementation of Random Forests for High Dimensional Data in C++ and R. J Stat Softw 77:1-17. <a href="https://doi.org/10.18637/jss.v077.i01" class="external-link">doi:10.18637/jss.v077.i01</a>
.</p></li>
<li><p>Schmid, M., Wright, M. N. &amp; Ziegler, A. (2016). On the use of Harrell's C for clinical risk prediction via random survival forests. Expert Syst Appl 63:450-459. <a href="https://doi.org/10.1016/j.eswa.2016.07.018" class="external-link">doi:10.1016/j.eswa.2016.07.018</a>
.</p></li>
<li><p>Wright, M. N., Dankowski, T. &amp; Ziegler, A. (2017). Unbiased split variable selection for random survival forests using maximally selected rank statistics. Stat Med. <a href="https://doi.org/10.1002/sim.7212" class="external-link">doi:10.1002/sim.7212</a>
.</p></li>
<li><p>Breiman, L. (2001). Random forests. Mach Learn, 45(1), 5-32. <a href="https://doi.org/10.1023/A%3A1010933404324" class="external-link">doi:10.1023/A:1010933404324</a>
.</p></li>
<li><p>Ishwaran, H., Kogalur, U. B., Blackstone, E. H., &amp; Lauer, M. S. (2008). Random survival forests. Ann Appl Stat 2:841-860. <a href="https://doi.org/10.1097/JTO.0b013e318233d835" class="external-link">doi:10.1097/JTO.0b013e318233d835</a>
.</p></li>
<li><p>Malley, J. D., Kruppa, J., Dasgupta, A., Malley, K. G., &amp; Ziegler, A. (2012). Probability machines: consistent probability estimation using nonparametric learning machines. Methods Inf Med 51:74-81. <a href="https://doi.org/10.3414/ME00-01-0052" class="external-link">doi:10.3414/ME00-01-0052</a>
.</p></li>
<li><p>Hastie, T., Tibshirani, R., Friedman, J. (2009). The Elements of Statistical Learning. Springer, New York. 2nd edition.</p></li>
<li><p>Geurts, P., Ernst, D., Wehenkel, L. (2006). Extremely randomized trees. Mach Learn 63:3-42. <a href="https://doi.org/10.1007/s10994-006-6226-1" class="external-link">doi:10.1007/s10994-006-6226-1</a>
.</p></li>
<li><p>Meinshausen (2006). Quantile Regression Forests. J Mach Learn Res 7:983-999. <a href="https://www.jmlr.org/papers/v7/meinshausen06a.html" class="external-link">https://www.jmlr.org/papers/v7/meinshausen06a.html</a>.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="predict.blockForest.html">predict.blockForest</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Marvin N. Wright</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bips-hb/blockForest" class="external-link">blockForest</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Standard Block Forest</span></span></span>
<span class="r-in"><span><span class="fu">blockForest</span><span class="op">(</span><span class="va">Species</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">iris</span>, </span></span>
<span class="r-in"><span>            blocks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>            mtry <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>            block.weights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.9</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>            block.method <span class="op">=</span> <span class="st">"BlockForest"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> blockForest result</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  blockForest(Species ~ ., iris, blocks = list(1:2, 3:4), mtry = c(1,      2), block.weights = c(0.1, 0.9), block.method = "BlockForest") </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Type:                             Classification </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of trees:                  500 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample size:                      150 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of independent variables:  4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Mtry:                             1 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Target node size:                 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variable importance mode:         none </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> OOB prediction error:             4.00 % </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Without blocks, grow standard random forest</span></span></span>
<span class="r-in"><span><span class="fu">blockForest</span><span class="op">(</span><span class="va">Species</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">iris</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> blockForest result</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  blockForest(Species ~ ., iris) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Type:                             Classification </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of trees:                  500 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample size:                      150 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of independent variables:  4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Mtry:                             2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Target node size:                 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variable importance mode:         none </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> OOB prediction error:             4.67 % </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marvin N. Wright.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

